pkg/cloud/instance.go

```diff
@@ -302,38 +312,163 @@ func (c *client) CheckLimits(
+
+	resp, err := c.cs.Address.ListPublicIpAddresses(params)
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to check free IPs for network %q", resolvedNet.Id)
+	}
+
+	for _, addr := range resp.PublicIpAddresses {
+		if addr.State == "Free" {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func (c *client) buildStaticIPEntry(ip, networkID string, resolvedNet *cloudstack.Network) (map[string]string, error) {
+	if err := c.validateIPInCIDR(ip, resolvedNet, networkID); err != nil {
+		return nil, err
+	}
+
+	if resolvedNet.Type == "Shared" {
+		isAvailable, err := c.isIpAvailableInNetwork(ip, networkID)
+		if err != nil {
+			return nil, err
+		}
+		if !isAvailable {
+			return nil, errors.Errorf("IP %q is already allocated in network %q or out of range", ip, networkID)
+		}
+	}
+
+	return map[string]string{
+		"networkid": networkID,
+		"ip":        ip,
+	}, nil
+}
+
+func (c *client) buildDynamicIPEntry(resolvedNet *cloudstack.Network) (map[string]string, error) {
+	if resolvedNet.Type == "Shared" {
+		freeIPExists, err := c.hasFreeIPInNetwork(resolvedNet)
+		if err != nil {
+			return nil, err
+		}
+		if !freeIPExists {
+			return nil, errors.Errorf("no free IPs available in network %q", resolvedNet.Id)
+		}
+	}
+
+	return map[string]string{
+		"networkid": resolvedNet.Id,
+	}, nil
+}
+
+func (c *client) resolveNetworkByName(name string) (*cloudstack.Network, error) {
 	net, count, err := c.cs.Network.GetNetworkByName(name, cloudstack.WithProject(c.user.Project.ID))
 	if err != nil {
-		return "", errors.Wrapf(err, "failed to look up network %q", name)
+		return nil, errors.Wrapf(err, "failed to look up network %q", name)
 	}
 	if count != 1 {
-		return "", errors.Errorf("expected 1 network named %q, but got %d", name, count)
+		return nil, errors.Errorf("expected 1 network named %q, but got %d", name, count)
 	}
-	return net.Id, nil
+	return net, nil
 }
 
 func (c *client) buildIPToNetworkList(csMachine *infrav1.CloudStackMachine) ([]map[string]string, error) {
-	ipToNetworkList := []map[string]string{}
+	var ipToNetworkList []map[string]string
 
 	for _, net := range csMachine.Spec.Networks {
-		networkID := net.ID
-		if networkID == "" {
-			var err error
-			networkID, err = c.resolveNetworkIDByName(net.Name)
+		networkID, resolvedNet, err := c.resolveNetworkReference(net)
+		if err != nil {
+			return nil, err
+		}
+
+		var entry map[string]string
+		if net.IP != "" {
+			entry, err = c.buildStaticIPEntry(net.IP, networkID, resolvedNet)
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			entry, err = c.buildDynamicIPEntry(resolvedNet)
 			if err != nil {
 				return nil, err
 			}
 		}
-		entry := map[string]string{"networkid": networkID}
-		if net.IP != "" {
-			entry["ip"] = net.IP
-		}
+
 		ipToNetworkList = append(ipToNetworkList, entry)
 	}
+
 	return ipToNetworkList, nil
 }
 
+func (c *client) resolveNetworkReference(net infrav1.NetworkSpec) (string, *cloudstack.Network, error) {
```

Comment:
```suggestion
func (c *client) resolveNetwork(net infrav1.NetworkSpec) (string, *cloudstack.Network, error) {
```
to maintain consistency